src/modules/emote_autocomplete/twitch/EmoteAutocomplete.jsx

```diff
@@ -0,0 +1,182 @@
+    EmoteCategories.BETTERTTV_GLOBAL,
+    EmoteCategories.BETTERTTV_PERSONAL,
+    EmoteCategories.FRANKERFACEZ_CHANNEL,
+    EmoteCategories.FRANKERFACEZ_GLOBAL,
+  ]);
+
+  const emoteSet = createTwitchEmoteSet(allEmotes);
+  const index = autocompleteEmotes.findIndex(({id}) => id === CUSTOM_SET_ID);
+
+  if (index === -1) {
+    autocompleteEmotes.push(emoteSet);
+  } else {
+    autocompleteEmotes[index] = emoteSet;
+  }
+
+  autocompleteEmoteProvider.forceUpdate();
+}
+
+let cleanup = null;
+export default class EmoteAutocomplete {
+  constructor() {
+    watcher.on('channel.updated', () => this.load());
+    settings.on(`changed.${SettingIds.EMOTE_AUTOCOMPLETE}`, () => this.load());
+  }
+
+  // on chat load, add emotes to twitch's autocomplete emote provider
+  // on autocomplete match render, replace image source of betterttv emotes with bttv cdn links
+
+  patchEmoteImage(image, isConnected) {
+    if (!isConnected) {
+      return;
+    }
+
+    const url = String(image.srcset);
+    const deseralizedEmote = deserializeEmoteFromURL(url);
+
+    if (deseralizedEmote == null) {
+      return;
+    }
+
+    const emote = emotes.getEligibleEmote(deseralizedEmote.code);
+    if (emote == null) {
+      return;
+    }
+
+    image.srcset = createSrcSet(emote.images);
+    image.src = emote.images['1x'];
+  }
+
+  unload() {
+    cleanup();
+    cleanup = null;
+
+    const autocompleteEmoteProvider = getAutocompleteEmoteProvider();
+    if (autocompleteEmoteProvider == null) {
+      return;
+    }
+
+    autocompleteEmoteProvider.props.emotes = autocompleteEmoteProvider.props.emotes.filter(
+      ({id}) => id !== CUSTOM_SET_ID
+    );
+
+    autocompleteEmoteProvider.forceUpdate();
+  }
+
+  load() {
+    if (cleanup != null) {
+      this.unload();
+    }
+
+    if (!settings.get(SettingIds.EMOTE_AUTOCOMPLETE)) {
+      return;
+    }
+
+    const emoteAutocompleteProvider = getAutocompleteEmoteProvider();
+    if (emoteAutocompleteProvider == null) {
+      return;
+    }
+
+    const oldComponentDidUpdate = emoteAutocompleteProvider.componentDidUpdate;
+    emoteAutocompleteProvider.componentDidUpdate = function componentDidUpdate(prevProps) {
+      if (prevProps.emotes !== this.props.emotes) {
+        injectEmoteSets();
+      }
+
+      if (oldComponentDidUpdate != null) {
+        oldComponentDidUpdate.call(this, ...prevProps);
+      }
+    };
+
+    const dirtyCallback = () => {
+      if (!emoteMenuViewStore.isLoaded()) {
+        emoteMenuViewStore.once('updated', () => injectEmoteSets());
+      } else {
+        injectEmoteSets();
+      }
+    };
+
+    dirtyCallback();
+
+    const storeCallbackCleanup = emoteMenuViewStore.on('dirty', dirtyCallback);
+    const patchImageCallbackCleanup = dom.on(AUTOCOMPLETE_MATCH_IMAGE_QUERY, this.patchEmoteImage);
+
+    cleanup = () => {
+      storeCallbackCleanup();
+      patchImageCallbackCleanup();
+      emoteAutocompleteProvider.componentDidUpdate = oldComponentDidUpdate;
```

Comment:
does this cleanup process actually work, or will this just break the react component if there is a method we stored? to my knowledge you can't restore this reference if it changed. see the send message patching for how you need to approach this: https://github.com/night/betterttv/blob/master/src/modules/send_message/index.js#L69-L81